### PR TITLE
Updating Calico manifests to Calico release 2.6.2

### PR DESCRIPTION
**Pull request by chrislovecnm**
Renamed the k8s-1.8 manifest to a k8s-1.7. This is required because of config
change that occurs between k8s 1.6 and k8s 1.7. This refactor will also
be re-used when Calico Kubernetes data source support is added to kops.
Updated bootstrapchannelbuilder with the new Calico version numbers.

The diffs for the k8s-1.6 version is pretty rough, tried to make it cleaner, but nada.

FIXES: https://github.com/kubernetes/kops/issues/3866
FIXES: https://github.com/kubernetes/kops/issues/3867

Line: https://github.com/kubernetes/kops/compare/master...chrislovecnm:calico-2.6-update?expand=1#diff-891cbc61587adb202b66b7c9bc6896daR209  is why Calico would not start on k8s 1.6 - thanks @caseydavenport 

 **TODO**
Testing K8s versions

- [x] test 1.5
- [x] test 1.6
- [x] test 1.7
- [x] test 1.8

/cc @blakebarnett @itajaja

### CODE REVIEW

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template [hunk 1]
```diff
@@ -16,10 +16,11 @@ data:
   calico_backend: "bird"
 
   # The CNI network configuration to install on each node.
+  # cniVersion should be 0.1.0 on k8s: https://github.com/projectcalico/calico/issues/742
   cni_network_config: |-
     {
       "name": "k8s-pod-network",
-      "cniVersion": "0.3.0",
+      "cniVersion": "0.1.0",
```
Comment (caseydavenport) [inline upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: Does this work?

I'd expect you to need to also remove the list part here.

For example:

```json
    {
       "name": "k8s-pod-network",
       "cniVersion": "0.1.0",
       "type": "calico",
       "etcd_endpoints": "__ETCD_ENDPOINTS__",
       "log_level": "info",
       "ipam": {
         "type": "calico-ipam"
       },
       "policy": {
         "type": "k8s",
         "k8s_api_root": "https://__KUBERNETES_SERVICE_HOST__:__KUBERNETES_SERVICE_PORT__",
         "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"
       },
       "kubernetes": {
         "kubeconfig": "/etc/cni/net.d/__KUBECONFIG_FILENAME__"
       }
     }
```

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template [hunk 2]
```diff
@@ -237,30 +249,75 @@ spec:
 
 ---
 
-# This manifest deploys the Calico policy controller on Kubernetes.
-# See https://github.com/projectcalico/k8s-policy
+# This deployment turns off the old "policy-controller". It should remain at 0 replicas, and then
```
Comment (caseydavenport) [inline upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: Not sure how upgrade works, but we'll only need this here if it's possible to upgrade from a v1.6 cluster with Calico v2.4 to a cluster with Calico v2.6.

It won't hurt either way.

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 1 above)
Comment (caseydavenport) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: Yep, like this one: https://github.com/kubernetes/kops/pull/3869/files#diff-aefa6a2023a7eddb4e0106cb7c132b96R19

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template [hunk 3]
```diff
@@ -69,7 +69,7 @@ spec:
         # container programs network policy and routes on each
         # host.
         - name: calico-node
-          image: quay.io/calico/node:v2.4.0
+          image: quay.io/calico/node:v2.6.2
```
Comment (caseydavenport) [inline upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template]: If you've managed to test this and show that it works, then I'm OK with it, but note that we haven't extensively tested Calico v2.6.0 on Kubernetes < v1.6, so it's sort of uncharted territory.

I'm not aware of a reason it won't work though.

Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template]: Should we upgrade then?  I think a CVE was fixed in Calico?

Comment (blakebarnett) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template]: I think we should stick with the most tested configuration, I assume a CVE would have been fixed in 2.4.x?

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 2 above)
Comment (blakebarnett) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: Yeah should probably remove it.

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 1 above)
Comment (blakebarnett) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: I think this got switched because of this: https://github.com/projectcalico/calico/issues/742 could you verify that's still the best version?

Comment (caseydavenport) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: It depends on the Kubernetes version in use.

Anything less than k8s v1.7 (so v1.5, v1.6) will need CNI version 0.1.0 with the config specified here.

For Kubernets v1.7+, CNI 0.3.0 with the chained CNI config is fine.

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template (hunk 3 above)
Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template]: We need to go to the new policy controller for kops and weave as well.  The old policy controller does not work :(

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 1 above)
Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: This is the template for 1.7, let me check the 1.7 template

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 2 above)
Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: @blakebarnett we need to leave it, as an upgrade will not work without it.  We do a kubectl apply, which will not delete old stuff.

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template (hunk 3 above)
Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template]: smoke tested 1.5.8 k8s with go-guest book ... worked like a champ.  Will upgrade and let Calico support or tell the users to upgrade clusters :)

**Comment by chrislovecnm** [comment]
/approve

@caseydavenport or @blakebarnett can I get a lgtm?

**Comment by chrislovecnm** [comment]
@justinsb I have not tested with gossip ... let me do that as well

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 2 above)
Comment (blakebarnett) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: no, I mean the new 2.6.2 version of `calico-kube-controllers` should be removed and we should run v2.4.x on k8s 1.6 with `calico-policy-controller` because it's the most widely tested configuration.

**Comment by chrislovecnm** [comment]
gossip and k8s 1.5 -> k8s 1.8 are happy now.  @blakebarnett @andrewsykim @geojaz PTAL

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 2 above)
Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: We cannot run `calico-policy-controller` on a weave cluster.  It does not work.

Comment (blakebarnett) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: of course not... :)  This is only part of the calico addon.

**Comment by chrislovecnm** [comment]
@caseydavenport so with the older version of Calico you cannot run a k8s 1.5 cluster with kops gossip.  Does Calico support version 2.6.2 with k8s version 1.5.x?

**Comment by blakebarnett** [comment]
I don't feel like forcing the upgrade to 2.6.2 for < k8s 1.7 is a good idea.  If the calico folks disagree then I'm happy.

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 1 above)
Comment (caseydavenport) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: > /networking.projectcalico.org/k8s-1.6.yaml.template

Looks like it's the 1.6 template?

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template (hunk 3 above)
Comment (caseydavenport) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/pre-k8s-1.6.yaml.template]: @chrislovecnm sounds good to me.

**Comment by caseydavenport** [comment]
> I don't feel like forcing the upgrade to 2.6.2 for < k8s 1.7 is a good idea. If the calico folks disagree then I'm happy.

@blakebarnett @chrislovecnm I don't fully understand the implications of not supporting the gossip DNS, but I agree that I'd feel somewhat uncomfortable forcing upgrades to Calico v2.6.2 on older Kubernetes versions (< v1.6). The only reason we should do this is if the gossip DNS is a hard requirement.

Like I said, it should "just work", but the Calico team doesn't generally support or test such a drastic mismatch of versions, so if there is an issue we'd likely just have to say "please upgrade to a modern kubernetes release".

TL;DR - I agree with @blakebarnett - we shouldn't force an upgrade of Calico on older k8s versions unless not doing so will break existing users.

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 1 above)
Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: @caseydavenport confused.  What am I missing?

**Comment by chrislovecnm** [comment]
@caseydavenport @blakebarnett I backed out the pre-1.5 changes.  We will need release notes about using gossip.  

@caseydavenport do I need more changes on the 1.6 manifest?

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template (hunk 1 above)
Comment (caseydavenport) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: It could be me that is confused, but AFAICT this is the k8s v1.6 manifest template, but the CNI config has plugin chaining in it which isn't supported until k8s v1.7.

If I'm right, then the config needs to be changed to look like the v1.5. config does.

Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: To be clear we need
```yaml
# This ConfigMap is used to configure a self-hosted Calico installation.
kind: ConfigMap
apiVersion: v1
metadata:
  name: calico-config
  namespace: kube-system
data:
  # The calico-etcd PetSet service IP:port
  etcd_endpoints: "{{ $cluster := index .EtcdClusters 0 -}}
                      {{- range $j, $member := $cluster.Members -}}
                          {{- if $j }},{{ end -}}
                          http://etcd-{{ $member.Name }}.internal.{{ ClusterName }}:4001
                      {{- end }}"

# Configure the Calico backend to use.
  calico_backend: "bird"

  # The CNI network configuration to install on each node.
  cni_network_config: |-
    {
        "name": "k8s-pod-network",
        "type": "calico",
        "etcd_endpoints": "__ETCD_ENDPOINTS__",
        "log_level": "info",
        "ipam": {
            "type": "calico-ipam"
        },
        "policy": {
            "type": "k8s",
             "k8s_api_root": "https://__KUBERNETES_SERVICE_HOST__:__KUBERNETES_SERVICE_PORT__",
             "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"
        },
        "kubernetes": {
            "kubeconfig": "/etc/cni/net.d/__KUBECONFIG_FILENAME__"
        }
    }
```
ConfigMap in the 1.6 manifest?

**Comment by chrislovecnm** [comment]
@blakebarnett @caseydavenport PTAL

/approve

If I get a `/lgtm` it will merge.  @justinsb is waiting on final review from you guys

File: upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template [hunk 4]
```diff
@@ -5,46 +5,35 @@ metadata:
...
-  # The calico-etcd PetSet service IP:port
+  # etcd servers
   etcd_endpoints: "{{ $cluster := index .EtcdClusters 0 -}}
                       {{- range $j, $member := $cluster.Members -}}
                           {{- if $j }},{{ end -}}
                           http://etcd-{{ $member.Name }}.internal.{{ ClusterName }}:4001
                       {{- end }}"
 
-  # Configure the Calico backend to use.
+# Configure the Calico backend to use.
```
Comment (caseydavenport) [inline upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: commend indentation is off :)

Comment (chrislovecnm) [reply upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template]: 😆 yes it is ... fixing.  After the push `/lgtm`?
